README.md
# Photopollution Calculator for Munster

> This program uses a mathematical model to predict photopollution based on population density for Munster, Ireland. 
> This work is part of [BT Young Scientist](https://btyoungscientist.com/) project, 'Is It Possible to Create a Mathematical Model to Predict Photopollution Based on Population Density in Munster'

## Usage

To run this program, you will need:

* [Python](https://www.python.org/) (must be python 3)
* [Pandas](https://pandas.pydata.org/) (can be installed with `pip install pandas`)

If you wish to use `data/solver.py` to develop the model, you will also need to install [Scipy](https://www.scipy.org/) and [Matplotlib](https://matplotlib.org/). 
This can be done with `pip install scipy matplotlib`.

This program can be used as a command line tool.
First, you will need to download the source code from github.

```bash
$ git clone https://github.com/conorcaseyc/Photopollution-Calculator
$ cd Photopollution-Calculator
```

Then you can run the program

```bash
$ python3 main.py town Limerick # Get the conditions for Limerick
$ python3 main.py pd 1500 # Get the conditions a given population density
```

To see the tool being used, have a look at the following recording.

[![asciicast](https://asciinema.org/a/8TDA52pkhfXEOAPplZqTqEM3D.svg)](https://asciinema.org/a/8TDA52pkhfXEOAPplZqTqEM3D)

If you have any problems, please leave an issue.
For general feedback, please [email the developer](mailto:16ccasey@student.kenmarecs.com).

## Information

> Where did the population density for the towns come from?

They came from the Irish 2016 Census, and as a result are provided by the Central Statistics Office.

> Why does the program crash when I enter a town name into the program?

This is most likely due to the fact that you entered a town that is not included in the Central Statistics Office database for population density. There is also a probability you typed "license", "quit", or "help" into the section where you are asked to input a town name. These options can only be accessed when you are prompted "Is the name of the town being entered:".
  
> How do I open the files in the Map Data folder

All these files can be opened using a program called, QGIS (previously known as Quantum GIS). QGIS is a free and open-source cross-platform desktop geographic information system (GIS) program that supports viewing, editing, and analysis of geospatial data. This program can be downloaded from [qgis.org](https://qgis.org/en/site/forusers/download.html)

## Contributing

Anyone is welcome to contribute to this project.

**Did you find a bug?**

* Ensure the bug was not already reported by searching on GitHub under [Issues](https://github.com/conorcaseyc/Photopollution-Calculator-for-Munster/issues).

* If you're unable to find an open issue addressing the problem, [open a new one](https://github.com/conorcaseyc/Photopollution-Calculator-for-Munster/issues/new). Be sure to include a **title and clear description**, as much relevant information as possible, and a **code sample** or an **executable test case** demonstrating the expected behavior that is not occurring.
* If possible, use the relevant bug report templates to create the issue. Simply copy the content of the appropriate template into a .rb file, make the necessary changes to demonstrate the issue, and **paste the content into the issue description**:
  * [**Bug Report** issues](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)
  * [**Feature Request** issues](https://github.com/rails/rails/blob/master/guides/bug_report_templates/action_controller_master.rb)

**Did you write a patch that fixes a bug?**

* Open a new GitHub pull request with the patch.
* Ensure the PR description clearly describes the problem and solution. Include the relevant issue number if applicable.
  


CHANGELOG.md
Photopollution Calculator for Munster Copyright (C) 2018  Conor Casey
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions. For further details, type `license'.

NB: Ensure the first letter of the town is capitalized!
Don't know how to use the program? Type "help" for support.
To exit the program, type 'quit'.

Updates:
v0.1.0: Initial beta version based on the mathematical model from the SciFest
project called "How Population Density Influences Light Pollution".
v0.1.1: Allowed population density figures to be inputted.
v0.5.0: The mathematical model is updated, and is now based on the 
BT Young Scientist project called "Is it Possible to Create a Mathematical Model
to Predict Photopollution Based on Population Density in Munster".
v0.5.1: Includes licensing agreement.
v0.5.2: Updated population density figures to include the figures from the 
2016 Census.
v0.5.3: See those bugs, no? Well, they are gone!
v0.5.4: Slight changes to the mathematical model.
v0.6.0: Slight changes to the grading system and mathematical model.
v0.6.1: Fixed TypeError when entering a text value for population density.
v0.6.2: See those bugs, no? Well, they are gone!
v0.6.3: Updated to support Windows and macOS alongside Linux.
v1.0.0: Worldwide Public Release on Github!
v1.0.1: Day 1 Bug Patch.
v1.1.0: Let's Get Py Crazy! update.
v1.1.1: Small Changes and Fixes
v1.2.0: Major Changes to Data Calculations
v1.5.0: Fin 2018 update
v1.5.1: Emergency Bug Fix Update

Current Edition: v1.5.1

main.py
import click
import math

CRED = '\033[91m'
CGREEN = '\033[92m'
CEND = '\033[0m'

@click.group()
def main():
  """
  This program uses a mathematical model to predict photopollution based on population density for Munster, Ireland.  
  
  This work is part of BT Young Scientist project, 'Is It Possible to Create a Mathematical Model to Predict Photopollution Based on Population Density in Munster'
  
  To predict the photopollution for a town, you can do:

  $ python3 main.py town Limerick

  To predict the photopollution from population density, you can do:

  $ python3 main.py pd 2500
  """
  pass

@main.command()
@click.argument('location', required=False)
def town(location):
  """
  Predict from a town name.
  """
  # This is done here to speed up other methods
  import pandas as pd

  # If no location is given, prompt.
  if not location:
    location = click.prompt("Please input the name of the town")

  pd_filename = 'data/population_density/populationdensitycensustowns.csv'
  pd_data = pd.read_csv(pd_filename)
  town = pd_data[pd_data['Towns'].str.contains(location, case=False)]

  if town.empty:
    return click.echo(CRED + 'No data on the town: {}'.format(location) + CEND)

  pd = float(town['PD'])
  print_conditions(pd)

@main.command()
@click.argument('population_density', required=False)
def pd(population_density):
  """
  Predict from population density.
  """

  while not population_density or not isfloat(population_density):
    if population_density:
      click.echo(CRED + 'Please input a valid population density' + CEND)

    population_density = click.prompt("Please input the population density")

  print_conditions(population_density)

@main.command()
def version():
  """
  Show the version history of the tool
  """
  with open('CHANGELOG.md', 'r') as fin:
    click.echo(fin.read())

@main.command()
def license():
  """
  Show the license of the tool
  """
  with open('LICENSE', 'r') as fin:
    click.echo(fin.read())


def get_conditions(lux):
  # Hard code in limerick's lux
  limerick_lux = 41.535984211999995

  # Get score as a percent
  score = (lux / limerick_lux) * 100

  message = '"Excellent Stargazing Conditions"'
  if score >= 80:
    message = "Terrible Stargazing Conditions"
  elif score >= 60:
    message = "Poor Stargazing Conditions"
  elif score >= 40:
    message = "Fair Stargazing Conditions"
  elif score >= 20:
    message = "Good Stargazing Conditions"

  return CGREEN + message + CEND, score	

def print_conditions(pd):
  lux = lux = 0.03510566 * float(pd) - 14.32414198
  conditions, score = get_conditions(lux)
  if score >= 0:
    		click.echo("Photopollution in this location is approximately """ + str(int(math.ceil(lux))) + " LUX, this should " + """correlate to """ + conditions)
  else:
  		click.echo("Oops, it appears we are getting a negative LUX value. Your population density is extremely low, therefore, this correlates to Excellent Stargazing Conditions.")


def isfloat(value):
  """Check if a string is a float"""
  try:
    float(value)
    return True
  except ValueError:
    return False

if __name__ == "__main__":
    main()

data/solver.py
from numpy import *
from scipy.interpolate import *
from matplotlib.pyplot import *
from scipy.stats import *

#The Data
population_density = array([21.5, 1205.4, 1351.3, 1636.4, 748.5, 1254, 1064, 1892, 3300, 1230.3, 658, 821.6, 1286.6, 1387.5, 1591.2, 1292, 1423.6, 1206.6, 1796.3, 768.5])
population = array ([557, 4979, 3593, 17140, 138, 1374, 946, 1346, 119320, 3765, 540, 3287, 25276, 9729, 94192, 1129, 6619, 23691, 14504, 2175])
lux = array([1, 21, 28, 38, 14, 34, 21, 50, 112, 25, 12, 14, 42, 28, 48, 24, 29, 20, 51, 12])
print(population_density)
print(population)
print(lux)

#Finds the formula to describe the correlation between population density and photopollution
p1 = polyfit(population_density,lux,1)
print(p1)

#Finds the correlation coefficient in the case of population density
slope,intercept,r_value,p_value,std_err = linregress(population_density,lux)
correlation = (pow(r_value,2))
print(correlation)
#Finds the correlation coefficient in the case of population
slope,intercept2,r_value2,p_value2,std_err2 = linregress(population,lux)
correlation2 = (pow(r_value2,2))
print(correlation2)

#Calculate Standard Deviation
mean = 0
for i in lux:
	mean += i
mean = mean / len(lux)
deviation_array = []
for i in lux:
	deviation = mean - i
	deviation_array.append(deviation)
sqaured_deviation = 0
for i in deviation_array:
	i = i ** 2
	sqaured_deviation += i
sqaured_deviation /= len(lux) - 1
standard_deviation = sqaured_deviation ** 0.5
print(standard_deviation)

#Plotting Graphs
#Mean LUX Values Scattering Plot
scatter(population_density,lux)
title("Mean LUX Values (All Sites)")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(population_density), poly1d(p1)(unique(population_density)))
text(0,100, "r = " + str(round(correlation, 3)))
show()

#Rest of Graphs
max_lux = array([1, 23, 32, 44, 16, 44, 23, 55, 125, 34, 16, 16, 52, 32, 52, 32, 34, 23, 58, 16])
min_lux = array([0, 18, 23, 32, 11, 23, 16, 44, 98, 16, 7, 11, 32, 23, 44, 16, 23, 16, 44, 7])
firsttrip_pd = array([21.5, 1205.4, 1351.3, 1636.4, 748.5, 1254, 1064, 1892, 3300, 1230.3])
secondtrip_pd = array ([658, 821.6, 1286.6, 1387.5, 1591.2, 1292, 1423.6, 1206.6, 1796.3, 768.5])
firsttrip_pp = array([1, 21, 28, 38, 14, 34, 21, 50, 112, 25])
secondtrip_pp = array([12, 14, 42, 28, 48, 24, 29, 20, 51, 12])
kerry_pd = array([21.5, 658, 1206.6, 1796.3, 768.5])
kerry_pp = array([1, 12, 20, 51, 12])
tipp_pd = array([1205.4, 1351.3, 1636.4])
tipp_pp = array([21, 28, 38])
water_pd = array([748.5, 1254, 1064])
water_pp = array([14, 34, 21])
cork_pd = array([1892, 3300, 1230.3])
cork_pp = array([50, 112, 25])
clare_pd = array([821.6, 1286.6, 1387.5])
clare_pp = array([14, 42, 28])
lim_pd = array([1591.2, 1292, 1423.6])
lim_pp = array([48, 24, 29])

#p(n)
p2 = polyfit(population_density,max_lux,1)
p3 = polyfit(population_density,min_lux,1)
p4 = polyfit(firsttrip_pd,firsttrip_pp,1)
p5 = polyfit(secondtrip_pd,secondtrip_pp,1)
p6 = polyfit(kerry_pd,kerry_pp,1)
p7 = polyfit(tipp_pd,tipp_pp,1)
p8 = polyfit(water_pd,water_pp,1)
p9 = polyfit(cork_pd,cork_pp,1)
p10 = polyfit(clare_pd,clare_pp,1)
p11 = polyfit(lim_pd,lim_pp,1)

#Max LUX Graph
slope,intercept,r_value,p_value,std_err = linregress(population_density,max_lux)
correlation = (pow(r_value,2))
scatter(population_density,max_lux)
title("Max LUX Values (All Sites)")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(population_density), poly1d(p2)(unique(population_density)))
text(0,120, "r = " + str(round(correlation, 3)))
show()

#Min LUX Graph
slope,intercept,r_value,p_value,std_err = linregress(population_density,min_lux)
correlation = (pow(r_value,2))
scatter(population_density,min_lux)
title("Minimum LUX Values (All Sites)")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(population_density), poly1d(p3)(unique(population_density)))
text(0,95, "r = " + str(round(correlation, 3)))
show()

#Mean LUX Graph (First Trip)
slope,intercept,r_value,p_value,std_err = linregress(firsttrip_pd,firsttrip_pp)
correlation = (pow(r_value,2))
scatter(firsttrip_pd,firsttrip_pp)
title("Mean LUX Values (First Trip)")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(firsttrip_pd), poly1d(p4)(unique(firsttrip_pd)))
text(0,100, "r = " + str(round(correlation, 3)))
show()

#Mean LUX Graph (Second Trip)
slope,intercept,r_value,p_value,std_err = linregress(secondtrip_pd,secondtrip_pp)
correlation = (pow(r_value,2))
scatter(secondtrip_pd,secondtrip_pp)
title("Mean LUX Values (Second Trip)")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(secondtrip_pd), poly1d(p5)(unique(secondtrip_pd)))
text(620,50, "r = " + str(round(correlation, 3)))
show()

#Kerry Graph
slope,intercept,r_value,p_value,std_err = linregress(kerry_pd,kerry_pp)
correlation = (pow(r_value,2))
scatter(kerry_pd,kerry_pp)
title("Kerry Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(kerry_pd), poly1d(p6)(unique(kerry_pd)))
text(0,50, "r = " + str(round(correlation, 3)))
show()
#Tipperary Graph
slope,intercept,r_value,p_value,std_err = linregress(tipp_pd,tipp_pp)
correlation = (pow(r_value,2))
scatter(tipp_pd,tipp_pp)
title("Tipperary Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(tipp_pd), poly1d(p7)(unique(tipp_pd)))
text(1200,37.5, "r = " + str(round(correlation, 3)))
show()
#Waterford Graph
slope,intercept,r_value,p_value,std_err = linregress(water_pd,water_pp)
correlation = (pow(r_value,2))
scatter(water_pd,water_pp)
title("Waterford Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(water_pd), poly1d(p8)(unique(water_pd)))
text(725,34, "r = " + str(round(correlation, 3)))
show()
#Cork Graph
slope,intercept,r_value,p_value,std_err = linregress(cork_pd,cork_pp)
correlation = (pow(r_value,2))
scatter(cork_pd,cork_pp)
title("Cork Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(cork_pd), poly1d(p9)(unique(cork_pd)))
text(1200,100, "r = " + str(round(correlation, 3)))
show()
#Clare Graph
slope,intercept,r_value,p_value,std_err = linregress(clare_pd,clare_pp)
correlation = (pow(r_value,2))
scatter(clare_pd,clare_pp)
title("Clare Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(clare_pd), poly1d(p10)(unique(clare_pd)))
text(800,40, "r = " + str(round(correlation, 3)))
show()
#Limerick Graph
slope,intercept,r_value,p_value,std_err = linregress(lim_pd,lim_pp)
correlation = (pow(r_value,2))
scatter(lim_pd,lim_pp)
title("Limerick Mean LUX Values")
xlabel("Population Density (people per km^2)")
ylabel("LUX Values")
plot(unique(lim_pd), poly1d(p5)(unique(lim_pd)))
text(1300,45, "r = " + str(round(correlation, 3)))
show()

main_3.py
changelog = open('CHANGELOG.md','r').read()
print(changelog)

def main():
	#Imports Population Density Data from the Central Statistics Office
	import pandas as pd
	import math
	pd_filename = 'data/population_density/populationdensitycensustowns.csv'
	pd_data = pd.read_csv(pd_filename)

	place_town = input("""
Is the name of the town being entered: """).lower()
	
	#Licensing Agreement
	if place_town == "license":
		license = open('LICENSE','r')
		license_read = license.read()
		print(license_read)
		license.close()
		main()
	#Force Closes Application	
	elif place_town == "quit":
		quit()
	elif place_town == "help":
		text = open('help.txt','r')
		text_read = text.read()
		print(text_read)
		text.close()
		main()					
	
	#Population Density or Town Input
	elif place_town == "no":
		place = input("""
Enter the Population Density: """)
		if place.isdigit() == False:
			print("It appears you have entered words instead of numbers, please try again.")
			main()
		else:
			print(place)
			user_input = float(place)
			
	#If A Town Is Entered
	elif place_town == "yes":	
		towns = input("""
Please input the name of the town: """)
		town = pd_data[pd_data.Towns.isin([towns])]
		town.reset_index(inplace = True, drop = True)	
		print(town) 
		user_input = float(town.PD)	
	else:
		print("Invalid Entry")
		main()	

	#Calculation of Light Pollution 		
	lux = 0.03510566 * user_input - 14.32414198

	#Get Limerick's Population Density In Order to Understand what the LUX Values Mean 
	limerick = pd_data[pd_data.Towns.isin(["Limerick City"])]
	limerick.reset_index(inplace = True, drop = True)
	limerick_pd = float(limerick.PD)
	limerick_lux = 0.03510566 * (limerick_pd) - 14.32414198

	#Understanding LUX Values
	calculations = (lux) / (limerick_lux) * 100
	def conditions():
		if calculations >= 80:
			return " Terrible Stargazing Conditions"
		elif calculations >= 60:
			return " Poor Stargazing Conditions"
		elif calculations >= 40:
			return " Fair Stargazing Conditions"
		elif calculations >= 20:
			return " Good Stargazing Conditions"
		elif calculations >= 0:
			return " Excellent Stargazing Conditions"			

	#Result/Output
	if calculations >= 0:
		print("""
Photopollution in this location is approximately """ + str(int(math.ceil(lux))) + " LUX, this should" + """
correlate to""" + conditions())
	elif calculations < 0:
		print("""
Oops, it appears we are getting a negative LUX value. Your population density
is extremely low, therefore, this correlates to Excellent Stargazing Conditions.""")

	#Restarts Program
	restart = input("""
Do You Want to Restart the Program: """).lower()
	if restart == "yes":
		print("""
		Restarting...""")
		main()
	else:
		quit()	
main()

#End of Code
